README.md
[![Latest Stable Version](https://poser.pugx.org/bitmotion/secure-downloads/v/stable)](https://packagist.org/packages/bitmotion/secure-downloads)
[![Total Downloads](https://poser.pugx.org/bitmotion/secure-downloads/downloads)](https://packagist.org/packages/bitmotion/secure-downloads)
[![Latest Unstable Version](https://poser.pugx.org/bitmotion/secure-downloads/v/unstable)](https://packagist.org/packages/bitmotion/secure-downloads)
[![License](https://poser.pugx.org/bitmotion/secure-downloads/license)](https://packagist.org/packages/bitmotion/secure-downloads)

# Secure Downloads (secure_downloads)

## Introduction

### What does it do?
In TYPO3, assets like PDFs, TGZs or JPGs etc. are normally just referenced by a URL e.g. to `fileadmin/…`. 
The file itself is delivered directly by the web server, and is therefore not part of the TYPO3 access control 
scheme. Files remain unprotected, since URLs can be re-used, emailed, Google-included or even guessed.

The extension **“Secure Download” (secure_downloads)** changes this behavior: Files will now be accessed through 
a eID Script script that honors TYPO3 access rights. The converted URL's will then look like this:

    https://www.your-domain.de/index.php?eID=tx_securedownloads&p=1009&u=1053&g=1&t=1561234595&hash=98ee3d7670a48cfef464d6aeff2aed1ed46f5f34&file=/fileadmin/content/protected/sample.doc

This works regardless of where the files come from, is not limited to special plugins etc.

Since in most cases you will not want to protect everything (which means that everything undergoes rather 
performance-consuming access right checking), Secure Download is highly configurable. You may choose:

* what directories to protect (e.g. you can include `typo3temp/` or not,)
* what file types to protect (do you want to protect JPGs or not? etc.)
* what domains are considered local

As a complementary measure, you will of course need to configure your web server not to deliver these things 
directly (e.g. using `.htaccess` settings).

## Administration

All configuration of this extension is made in the Extension Manager.

### Extension configurations
<table>
	<tr>
		<th>Tab</th>
		<th>Option</th>
		<th>Type</th>
		<th>Description</th>
	</tr>
	<tr>
		<td>Basic</td>
		<td>cachetimeadd</td>
		<td>Integer</td>
		<td>The secure link is only valid for a limited time, which is calculated from the cache time that is used for the page that carries the link plus this value (in seconds). Default is 3600 sec (= 1 h).</td>
	</tr>
	<tr>
		<td></td>
		<td>linkFormat</td>
		<td>String</td>
		<td>The Format for the link generated by secure_downloads. You can change this if you apply appropriate mod_rewrite rules in your .htaccess file.</td>
	</tr>
	<tr>
		<td></td>
		<td>enableGroupCheck</td>
		<td>boolesch</td>
		<td>Allows forwarding a secure download link to others, who can access that file if they have at least one front-end user group in common. Enabling this makes the checks LESS restrictive!</td>
	</tr>
	<tr>
		<td></td>
		<td>groupCheckDirs</td>
		<td>String</td>
		<td>A list of directories for the less restrictive group check, separated by a pipe (|). Leave empty if you want to enable the group check for all directories.</td>
	</tr>
	<tr>
		<td></td>
		<td>excludeGroups</td>
		<td>String</td>
		<td>A comma separated list of groups that are excluded from the group check feature (if enabled).</td>
	</tr>
	<tr>
		<td></td>
		<td>domain</td>
		<td>String</td>
		<td>A domain prefix before the secured directories part (e.g. http://www.host.com/). Not needed for internal (relative) links.</td>
	</tr>
	<tr>
		<td></td>
		<td>securedDirs</td>
		<td>String</td>
		<td>What directories shall be protected? Multiple directory patterns can be separated by a pipe (|).</td>
	</tr>
	<tr>
		<td></td>
		<td>securedFiletypes</td>
		<td>String</td>
		<td>What file types (read: extensions) shall be protected? Multiple file extension patterns can be separated by a pipe (|).</td>
	</tr>
	<tr>
		<th></th>
		<th></th>
		<th></th>
		<th></th>
	</tr>
	<tr>
		<td>Filedelivery</td>
		<td>apacheDelivery</td>
		<td>boolesch</td>
		<td>Feature to deliver files through Apache not PHP (Beta)</td>
	</tr>
	<tr>
		<td></td>
		<td>forcedownload</td>
		<td>boolesch</td>
		<td>Force Download of specified filetypes</td>
	</tr>
	<tr>
		<td></td>
		<td>forcedownloadtype</td>
		<td>String</td>
		<td>A list of filetypes that should not be opened inline in a browser, seperated by a pipe. Only used if forcedownload is true.</td>
	</tr>
	<tr>
		<td></td>
		<td>additionalMimeTypes</td>
		<td>String</td>
		<td>Comma separated list of additional MIME types (file extension / mime type pairs, in which file extension and MIME type is separated by a pipe symbol). Can be used to override existing MIME type settings of the extension as well.</td>
	</tr>
	<tr>
		<td></td>
		<td>outputFunction</td>
		<td>String</td>
		<td>PHP function which is used to deliver the secured files (one of readfile, readfile_chunked, fpassthru).</td>
	</tr>
	<tr>
		<td></td>
		<td>outputChunkSize</td>
		<td>Integer</td>
		<td>Size of chunks to be delivered in one go (only relevant if "readfile_chunked" is used) (default: 1048576 Bytes)</td>
	</tr>
	<tr>
		<th></th>
		<th></th>
		<th></th>
		<th></th>
	</tr>
	<tr>
		<td>Module</td>
		<td>log</td>
		<td>boolesch</td>
		<td>Log each file access which will be shown in a Module 'Download traffic'</td>
	</tr>
	<tr>
		<th></th>
		<th></th>
		<th></th>
		<th></th>
	</tr>
	<tr>
		<td>Debug</td>
		<td>debug</td>
		<td>String</td>
		<td>Output each tag that maches the regex and will be replaced (default: 0)</td>
	</tr>
</table>

### Details on configuration evaluation
* The config fields *"securedDirs"*, *"filetype"* and *"domain"* allow regular expressions, but with limitations because some 
characters (slash, backslash, dot, blank) are automatically quoted for your convenience.
* For *"filetype"* (meaning actually the file extension), all upper/lowercase combinations are automatically included 
(e.g. "gif" would also cover "giF".)
* The `|` stands for "OR" .

#### Regex examples for securedDirs:

If for example you need to secure `fileadmin` and `typo3temp`, but not `uploads`:


    fileadmin|typo3temp


To secure everything under `fileadmin/secure` or `typo3temp`, you need to write:

    fileadmin/secure|typo3temp

You also can group some elements with regular expression, but you should be carefull with grouping because the complex 
regex in the extension does not work if some other matches were output by the regex. For grouping `( )` is used, 
but in our case you need to exclude the result of this `( )`. This can be done with `(?: )`. 

For example we need to find all under `fileadmin/secure1`, `fileadmin/secure2`, `fileadmin/secure3` or `typo3temp`:

    fileadmin/secure(?:1|2|3)|typo3temp

For the same need you can use also `[ ]` but all chars between `[ ]` are allowed here:

    fileadmin/secure[123]|typo3temp

If you need to exclude some subfolders in a secured directory you can do this by `(?! )`. For example like:

    (?!fileadmin/unsecured)fileadmin

More information can be found here: http://www.regular-expressions.info

### Access configuration

You also need to secure all the directories and filetypes by your server configuration. This can be done with 
`.htaccess` files. Some example `.htaccess` files can be found in the `Resources/Private/Examples` folder.

**Note:** This extension cannot secure links to files that you include in your CSS file. For example you can secure 
`/fileadmin` with the default `.htaccess_deny` file by putting the file in `/fileadmin`. You can allow 
`/fileadmin/templates/` with the default `.htaccess_allow` file by putting this file to `/fileadmin/template/`.


### Backend Module
Since version 1.0.0 this extensions contains a backend module. With this module you can show the summerized download 
traffic by a given timeframe for all users or by a frontend user. Since version 1.2.0 you will see all files and 
filesizes downloaded by specified users.

## Known Limitations / ToDo
* Get a hook for showpic.php – currently, there is none (thus xclassing necessary).
* Support for links from one PDF to another.
* Allow optional advanced mode for full Regex support.
* Images in Direct Mail newsletters, doesn't work correct with secure_dl :(
* Add a reST documentation.


ChangeLog
2013-12-03 Dietrich Heise <typo3-ext(at)bitmotion.de>, Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* Namespaces, Stucture, PSR-0
	* Raised extension version to 1.8.0

2013-07-19 Dietrich Heise <typo3-ext(at)bitmotion.de>, Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* Rewritten for 6.x
	* Works with TYPO3 4.5 and greater
	* Bug #50629 and #52441 fixed BE thumbnails broken
	* Bug #52003 fixed Downloads broken in 4.5.30

2013-07-19 Dietrich Heise <typo3-ext(at)bitmotion.de>
	* First Version with TYPO3 6.x Support
	* Change the order of GET params in the default settings, so &file=/path/example.jpg is at the end of the url,
	  because some javascript functions check for the file extension (e.g. jQuery Colorbox)

2012-08-15 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* Enable link generation with user groups (Thanks to punkt.de for sponsoring).

2012-08-14 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* Improve performance of file extension regex (Thanks to Sebastian Kurfürst).

2011-06-14 Dietrich Heise  <typo3-ext(at)bitmotion.de>
	* Add support for HTML5 Tag <source src=""> (Thanks to Achim Schmeer for sponsoring this).

2011-06-14 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* Add config option to disable content parsing (Thanks to Benjamin Mack).

2011-03-14 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* Added possibility to configure the link format, to be able to have nicer links, when also adding
	  an appropriate RewriteRule in the .htaccess (Thanks to Karl-Ulrich Bamberg for the initial idea and patch).
	* Raised extension version to 1.5.0

2011-03-08 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* Removed deprecated functions (thanks to Daniel Minder)

2011-02-25 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* Added workaround for UTF-8 filenames on Windows systems. This is a PHP bug and
	  will only work for special characters in filenames, which are present in latin1 charset.
	  See http://bugs.php.net/bug.php?id=46990 for details. (thanks to Freddy Kühne for reporting)

2010-11-25 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* Possibility to add mime types in extension manager configuration (thanks to Ingmar Schlecht)

2010-11-18 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* Add MS-Office and OpenOffice mime types (thanks to Christian Bülter)
	* add selected mime type to hook params to enable the hook to modify them
	* use 3tlib_div::makeInstance instead of new in tx_nawsecuredl_output

2010-11-10 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* version 1.4.3
	* use t3lib_div::hmac instead of php native function to make use of a fallback
	  if hash_hmac is not present (thanks to Jan Wulff)
	* Fix XCLASS inclusion in class.tx_nawsecuredl_output.php (thanks to Jan Wulff)

2010-06-11 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* version 1.4.2
	* fix urlencode regression bug introduced in 1.4.1 (Thanks to Felix Nagel)

2010-04-29 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* version 1.4.1
	* add hook requested by Marco Huber
	* add new possibilty to output the files in chunks
	* make file output method configurable
	* send Content-Length: header to enable progress indicators to function
	* downloads are now logged every time, but with 0 bytes, if the download process is aborted
	* log only the downloaded bytes if readfile_chunked is used as output method

2010-01-20 Dietrich Heise  <typo3-ext(at)bitmotion.de>
	* add .htm/.html files to mime types

2010-01-20 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* version 1.3.0
	* use hmac function to generate the hash
	* table view for backend module
	* added check, if readfile function is available
	* improve logging, to not log, if download is aborted

2009-08-12 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* version 1.2.0
	* table view for backend module

2009-07-07 Helmut Hummel  <typo3-ext(at)bitmotion.de>
	* version 1.0.0
	* added logging of filesize
	* added backend module for admins to display traffic usage by time and user

2009-05-29 Dietrich Heise  <typo3-ext(at)bitmotion.de>
	* fixed hooks

2008-11-28 Dietrich Heise  <typo3-ext(at)bitmotion.de>
	* added parsing for JavaScript window.open (vHWin=window.open)

2008-06-17 Dietrich Heise  <typo3-ext(at)bitmotion.de>
	* added some Hooks, feel free to use ;)

2007-10-17 Dietrich Heise  <typo3-ext(at)bitmotion.de>
	* fixed the problem with downloads, previews in the Backend,
	thanks to Bedrijvenweb Ontwikkeling

2007-06-20 Dietrich Heise  <heise@naw.de>
	* version 0.2.2
	* various changes with the eID file
	* fixed &amp; quoted char for the generated URL for each file
	* enabled logging feature for file access (no output of the logging at the moment)

2007-01-23 Dietrich Heise  <heise@naw.de>
	* Fixed problem with getimagesize()

2006-02-23 Dietrich Heise  <heise@naw.de>
	* Fixed "\" bug on Windows Platforms

2005-11-07 Dietrich Heise  <heise@naw.de>

	* Initial release


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "naw_securedl".
 *
 * Auto generated 20-08-2013 16:44
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Secure Downloads',
	'description' => '"Secure Download": Apply TYPO3 access rights to ALL file assets (PDFs, TGZs or JPGs etc. - configurable) - protect them from direct access.',
	'category' => 'fe',
	'shy' => 0,
	'version' => '1.8.0',
	'dependencies' => 'cms',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => 'modLog',
	'state' => 'stable',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'Dietrich Heise, Helmut Hummel',
	'author_email' => 'typo3-ext(at)bitmotion.de',
	'author_company' => '<a href="http://www.bitmotion.de" target="_blank">bitmotion.de</a>',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
			'cms' => '',
			'php' => '5.3.0-5.5.99',
			'typo3' => '4.5.0-6.2.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:27:{s:9:"ChangeLog";s:4:"6f09";s:28:"class.tx_nawsecuredl_log.php";s:4:"25c5";s:31:"class.tx_nawsecuredl_output.php";s:4:"4f10";s:21:"class.ux_fileList.inc";s:4:"05e0";s:29:"class.ux_SC_tslib_showpic.php";s:4:"81d0";s:16:"ext_autoload.php";s:4:"2f51";s:21:"ext_conf_template.txt";s:4:"6714";s:12:"ext_icon.gif";s:4:"2dbb";s:17:"ext_localconf.php";s:4:"831e";s:14:"ext_tables.php";s:4:"9fc5";s:14:"ext_tables.sql";s:4:"ec93";s:8:"TODO.txt";s:4:"014d";s:37:"Classes/Driver/Xclass/LocalDriver.php";s:4:"f845";s:35:"Classes/Service/SecureDownloadService.php";s:4:"5bce";s:41:"Classes/Service/Tx_NawSecuredl_Service_SecureDownloadService.php";s:4:"dae3";s:14:"doc/manual.sxw";s:4:"c85e";s:37:"modLog/class.tx_nawsecuredl_table.php";s:4:"f26e";s:15:"modLog/conf.php";s:4:"9c2f";s:16:"modLog/index.php";s:4:"7cb0";s:20:"modLog/locallang.xml";s:4:"0b71";s:24:"modLog/locallang_mod.xml";s:4:"84d1";s:21:"modLog/moduleicon.gif";s:4:"691d";s:14:"res/_.htaccess";s:4:"e99d";s:20:"res/_.htaccess_allow";s:4:"5057";s:19:"res/_.htaccess_deny";s:4:"e99d";s:29:"tests/SecureDownloadServiceTest.php";s:4:"b077";s:35:"tests/tx_nawsecuredl_outputTest.php";s:4:"94bd";}',
	'suggests' => array(
	),
);

?>
